File: API/Entities/Basket.cs
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations.Schema;
using System.Linq;
using System.Threading.Tasks;

namespace API.Entities
{

    public class Basket
    {
        public int Id { get; set; }
        public required string BasketId { get; set; }
        public List<BasketItem> Items { get; set; } = [];
        public string? ClientSecret { get; set; }
        public string? PaymentIntentId { get; set; }
        public void AddItem(Product product, int quantity)
        {
            if (product == null) ArgumentNullException.ThrowIfNull(product);
            if (quantity <= 0) throw new ArgumentOutOfRangeException("Quantity should be greater than zero",
             nameof(quantity));

            var existingItem = FindItem(product.Id);

            if (existingItem == null)
            {
                Items.Add(new BasketItem
                {
                    Product = product,
                    Quantity = quantity
                });
            }
            else
            {
                existingItem.Quantity += quantity;
            }


        }
        public void RemoveItem(int productId, int quantity)
        {
            if (quantity < 0) throw new ArgumentOutOfRangeException("Quantity should be greater than zero",
             nameof(quantity));
            var item = FindItem(productId);
            if (item == null) return;
            item.Quantity -= quantity;
            if (item.Quantity <= 0) Items.Remove(item);

        }

        public BasketItem? FindItem(int productId)
        {
            return Items.FirstOrDefault(item => item.ProductId == productId);
        }
    }


}

File: client/src/app/models/basket.ts
import type { Product } from "./product";

export type Basket = {
  basketId: string;
  items: Item[];
  clientSecret?: string;
  paymentIntentId?: string;
};

export class Item {
  constructor(product: Product, quantity: number) {
    this.productId = product.id;
    this.name = product.name;
    this.price = product.price;
    this.pictureUrl = product.pictureUrl;
    this.type = product.type;
    this.brand = product.brand;
    this.quantity = quantity;
  }
  productId: number;
  name: string;
  price: number;
  pictureUrl?: string;
  type: string;
  brand: string;
  quantity: number;
}
